test/parallel/test-stream-pipeline.js

```diff
@@ -0,0 +1,315 @@
+    const req = http.request({
+      port: server.address().port
+    });
+
+    req.end();
+    req.on('response', (res) => {
+      const buf = [];
+      res.on('data', (data) => buf.push(data));
+      res.on('end', common.mustCall(() => {
+        assert.deepStrictEqual(
+          Buffer.concat(buf),
+          Buffer.from('hello')
+        );
+        server.close();
+      }));
+    });
+  });
+}
+
+{
+  const server = http.createServer((req, res) => {
+    const rs = new Readable({
+      read() {
+        rs.push('hello');
+      },
+      destroy: common.mustCall((err, cb) => {
+        // prevents fd leaks by destroying http pipelines
+        cb();
+      })
+    });
+
+    pipeline(rs, res);
+  });
+
+  server.listen(0, () => {
+    const req = http.request({
+      port: server.address().port
+    });
+
+    req.end();
+    req.on('response', (res) => {
+      setImmediate(() => {
+        res.destroy();
+        server.close();
+      });
+    });
+  });
+}
+
+{
+  const server = http.createServer((req, res) => {
+    const rs = new Readable({
+      read() {
+        rs.push('hello');
+      },
+      destroy: common.mustCall((err, cb) => {
+        cb();
+      })
+    });
+
+    pipeline(rs, res);
+  });
+
+  let cnt = 10;
+
+  const badSink = new Writable({
+    write(data, enc, cb) {
+      cnt--;
+      if (cnt === 0) cb(new Error('kaboom'));
+      else cb();
+    }
+  });
+
+  server.listen(0, () => {
+    const req = http.request({
+      port: server.address().port
+    });
+
+    req.end();
+    req.on('response', (res) => {
+      pipeline(res, badSink, common.mustCall((err) => {
+        assert.deepStrictEqual(err, new Error('kaboom'));
+        server.close();
+      }));
+    });
+  });
+}
+
+{
+  const server = http.createServer((req, res) => {
+    pipeline(req, res, common.mustCall());
+  });
+
+  server.listen(0, () => {
+    const req = http.request({
+      port: server.address().port
+    });
+
+    const rs = new Readable({
+      read() {
+        rs.push('hello');
+      }
+    });
+
+    pipeline(rs, req, common.mustCall(() => {
+      server.close();
+    }));
+
+    req.on('response', (res) => {
+      let cnt = 10;
+      res.on('data', () => {
+        cnt--;
+        if (cnt === 0) rs.destroy();
+      });
+    });
+  });
+}
+
+{
+  const server = http2.createServer((req, res) => {
+    pipeline(req, res, common.mustCall());
+  });
+
+  server.listen(0, () => {
+    const url = `http://localhost:${server.address().port}`;
+    const client = http2.connect(url);
+    const req = client.request({ ':method': 'POST' });
+
+    const rs = new Readable({
+      read() {
+        rs.push('hello');
+      }
+    });
+
+    pipeline(rs, req, common.mustCall((err) => {
+      // TODO: this is working around an http2 bug
+      // where the client keeps the event loop going
+      // (replacing the rs.destroy() with req.end()
+      // exits it so seems to be a destroy bug there
+      client.unref();
```

Comment:
will be investigating this later on this afternoon :-)

Reply:
Testing this with other `Readable` instances shows this working. For instance, replace the custom `Readable` above with `const rs = fs.createReadStream(__filename)` and the test appears to work just fine.

Reply:
Yep, just verified, commenting out the `client.unref()` and replacing the custom `Readable` with `fs.createReadStream()` and adjusting the `cnt` below a bit to account for the difference in the number of `data` events allows this test to pass and the `Http2Stream` shuts down normally. So the issue *appears* to be specific to the custom `Readable`?? /cc @mcollina

Reply:
@jasnell hmm, i don't understand how the Readable should affect this. it's just the trigger of the destruction of the pipeline. could it be that there is an http2 issue if a bunch of writes happen *while* the stream is destroyed?

Reply:
Not that I'm aware of. At this point I'm a bit unsure what could be happening. I've tried a number of variations but can only reproduce it with the custom `Readable`. I've asked @mcollina to take a look. He said he will take a look tomorrow.

Reply:
@mcollina... Please don't forget to take a look at this one :)